packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol

```diff
@@ -158,4 +735,366 @@ library ConnextUtils {
+    _args.home.dispatch(_args.xCallArgs.params.destinationDomain, _args.remote, message);
+
+    return (
+      transferId,
+      message,
+      xCalledEventArgs({
+        transactingAssetId: transactingAssetId,
+        amount: amount,
+        bridgedAmt: bridgedAmt,
+        bridged: bridged
+      })
+    );
+  }
+
+  /**
+   * @notice Calculates a transferId base son `execute` arguments
+   * @dev Need this to prevent stack too deep
+   */
+  function _getTransferId(ExecuteLibArgs calldata _args) private view returns (bytes32) {
+    (uint32 tokenDomain, bytes32 tokenId) = _args.tokenRegistry.getTokenId(_args.executeArgs.local);
+
+    return
+      keccak256(
+        abi.encode(
+          _args.executeArgs.nonce,
+          _args.executeArgs.params,
+          _args.executeArgs.originSender,
+          tokenId,
+          tokenDomain,
+          _args.executeArgs.amount
+        )
+      );
+  }
+
+  /**
+   * @notice Calculates a transferId base son `xcall` arguments
+   * @dev Need this to prevent stack too deep
+   */
+  function _getTransferId(xCallLibArgs calldata _args, ConnextMessage.TokenId memory _canonical)
+    private
+    view
+    returns (bytes32)
+  {
+    return
+      keccak256(
+        abi.encode(
+          _args.nonce,
+          _args.xCallArgs.params,
+          msg.sender,
+          _canonical.id,
+          _canonical.domain,
+          _args.xCallArgs.amount
+        )
+      );
+  }
+
+  /**
+   * @notice Formats a nomad message generated by `xcall`
+   * @dev Need this to prevent stack too deep
+   */
+  function _formatMessage(
+    xCallLibArgs calldata _args,
+    address _asset,
+    bytes32 _transferId,
+    uint256 _amount
+  ) private returns (bytes memory) {
+    // get token
+    IBridgeToken token = IBridgeToken(_asset);
+
+    // declare details
+    bytes32 detailsHash;
+
+    if (_args.tokenRegistry.isLocalOrigin(_asset)) {
+      // TODO: do we want to store a mapping of custodied token balances here?
+
+      // token is local, custody token on this chain
+      // query token contract for details and calculate detailsHash
+      detailsHash = ConnextMessage.formatDetailsHash(token.name(), token.symbol(), token.decimals());
+    } else {
+      // if the token originates on a remote chain,
+      // burn the representation tokens on this chain
+      if (_amount > 0) {
+        token.burn(msg.sender, _amount);
+      }
+      detailsHash = token.detailsHash();
+    }
+
+    // format action
+    bytes29 action = ConnextMessage.formatTransfer(
+      TypeCasts.addressToBytes32(_args.xCallArgs.params.to),
+      _amount,
+      detailsHash,
+      _transferId
+    );
+
+    // get the tokenID
+    (uint32 domain, bytes32 id) = _args.tokenRegistry.getTokenId(_asset);
+
+    // format token id
+    bytes29 tokenId = ConnextMessage.formatTokenId(domain, id);
+
+    // send message
+    return ConnextMessage.formatMessage(tokenId, action);
+  }
+
+  function _handleExecuteTransaction(
```

Comment:
add a little function docstring pls, just with the `@notice base son` and `@dev Need this to prevent stack too deep` - that's actually an important clarifier imo